Use `property()` & implement Abstract Base Class `PangoObject`

cAttte opened the pull request:
hey! this PR closes #44. i ended up going with `property()` since i completely agree that it ties the methods better, _buuut_ i decided to keep `@property` for read-only properties, since otherwise it's unnecessarily verbose.

while writing this simple-ish PR, though, i realized that there was a lot of repetition between every single class, as they all need to interact with the underlying C APIs and pointers... so i created an abstract base class: `PangoObject`. this class implements the following methods/members, which removes a lot of repetition:

- `__init__()`
- `pointer`
- `_init_pointer()`
- `from_pointer()`
- `copy()` / `_copy_()` / `__deepcopy_()`
- `__eq__()`
- `__repr__()`

it works by utilizing these abstract properties, which are supposed to be provided by its subclasses:

- `_INIT_METHOD` -- a CFFI function that initializes the object (i.e. `pango.pango_..._new()`), alternatively, `ffi.new`
- `_INIT_CLASS` -- only if the method is set to `ffi.new`, a string representing the classname to initialize (e.g. `"PangoLayout")`
- `_GC_METHOD` -- a CFFI function that frees the object (i.e. `pango.pango_..._free()`)
- `_COPY_METHOD` -- a CFFI function that copies the object (i.e. `pango.pango_..._copy()`)
- `_EQ_METHOD` -- a CFFI function that compares the object with another (i.e. `pango.pango_..._equals()`)

all of these are optional, and every base method is enabled or modified depending on these (i.e., `__init__()` will error if there's no `_INIT_...` properties; `gc: bool` will be disabled if there's no `_GC_METHOD`; comparisons (`=`) will directly use pointers if an `_EQ_METHOD` isn't provided, etc.) of course, these base methods can still be overwritten like in the case of `Color#__eq__()`.

i'm sorry if this is kind of a sudden change, but it would've been pointless to make a PR that basically rewrites every class, to then make another one that rewrites them over again, and the current structure really needed a base class like this one. i believe the overall concept is good, but please let me know if there are any implementation details you'd like to change.

---

i updated all of the tests to conform to the changes and they should all be passing, but for some reason i'm geting seemingly random memory management errors (`Access violation` / `Segmentation fault`) _only_ when running the test suite (if i run the same functions from a simple test file nothing happens), and the errors happen on code that's basically untouched (`get_iter()` -> `pango_layout_get_iter`; `apply_markup()` -> `pango_layout_set_markup`). i'll try to see what's causing it because i'm not even sure if it has to do with my machine or what.

leifgehrmann commented:
First of all, great work! I do like the PangoObject refactoring.

I've ran the GitHub Actions just to see if the tests would run. Looks like there's a few issues that need to be fixed.

* Annoyingly the default installation of Ubuntu in GitHub actions uses a version of Pango older than 1.46.0. That version of Pango doesn't appear to have `pango_attr_list_equal`, which means the CFFI will raise an `AttributeError` when accessed at runtime. By moving that reference to the top of the file in `_EQ_METHOD = pango.pango_attr_list_equal`, it now raises an AttributeError whenever the pangocffi package is imported (Because `Attribute` is imported by `pangocffi/__init__.py`). Two workarounds I can think of at the moment are:
  * Upgrade the version of Pango used in the GitHub action, and drop support for version of Pango <1.46.0 (Though 1.46.0 was released only 2 years ago, so forcing an upgrade to use this package maybe isn't a good idea)
  * Remove `_EQ_METHOD` for `Attribute` and instead override the `__eq__` method directly. That way, the AttributeError will only be raised when `==` is used. (and the tests will pass because they take that into account)
* A few linting issues in `pango_object.py` and `test_color.py`.

(On another note, I really should split linting and tests into two different GitHub Action workflows)

I've also checked out the code on my machine as well and I can replicate the segfault errors you mentioned. I've not looked too closely at the new pointer logic, but I'm only guessing it's because a PangoObject is being freed when it's not supposed to be.

Unrelated to the SegFault stuff, I also noticed you fixed a typo made by a contributor in the `color.py` class. Specifically, in the `__eq__` method, where `and self.red == self.red` should have been `and self.blue == other.blue`. Thanks for spotting that! But strangely, when I run `pytest tests/functional/test_color.py::TestColor::test_copy`, the assertion appears to fail because apparently the property `blue` is `0`, instead of `5`... Curious if you get the same issue? I wonder if `pango_color_copy` is broken. For reference, I'm using Pango v1.50.10.

cAttte commented:
yeah, i saw the `attr_list_equal` error in the Actions and that's why i removed `_EQ_METHOD` from `AttrList`, but i guess overriding the method directly like you say is a better option, so i'll do that in a second. being that Pango v1 is over 20 years old, yeah, we should probably update it, even if not to 1.46.

i'll try to see if i can figure out why the segfault is happening, even though i don't have a lot of experience with C haha; the pointer logic really shouldn't be any different from the pre-existing classes, but it's probably a small detail that i missed. i'll also fix the linting issues; i ran `make lint` and fixed everything but i guess i must've changed something without realizing!

i noticed the weird error when copying a Color, but i forgot to mention it. it does seem like `pango_color_copy` might be broken, because i remember testing the FFI function directly and the object i got back was wrong. i'll try doing some more experimenting though. if it somehow does end up being broken we could just manually copy the RGB attributes to a new pointer? (and file an issue with pango!).

i also see another test case that asserts that a NotImplementedError is raised in `Color(...) == 1`, i assume i.e. whenever a Color is compared to a non-object... i had changed `PangoObject#__eq__()` so that it wouldn't error and simply return False; should i revert the change or remove the failing test case?

cAttte commented:
some updates:

- i overwrote `__eq__()`, and i changed it so it wouldn't error on an old version; just check if the comparison method exists, and if it doesn't just compare the pointers (we should probably somehow make it so they compare the attributes instead, like `Color.__eq__()` does)
- i still don't know what causes the Color copy issue, but the following code reproduces it:
```py
from pangocffi import ffi, pango

color = ffi.new("PangoColor *")
color.red = 55
color.green = 155
color.blue = 255

copy = pango.pango_color_copy(color)
print(color.blue, copy.blue) # 255, 0

spec = ffi.new("char[]", "#111".encode("utf8"))
```

in addition to that, other methods like `pango_color_parse` also return a `blue` component of `0`, but if we generate a string representation of the color, it looks like it's still present?:

```py
spec = ffi.new("char[]", "#102030".encode("utf8"))
new_color = ffi.new("PangoColor *")
pango.pango_color_parse(new_color, spec)
print(new_color.blue, ffi.string(pango.pango_color_to_string(new_color))) # 0, b"#101020203030"
```

- the segmentation fault seems to only ever present in `Layout`; specifically:
  - `apply_markup()` (2nd line); 
  - `get_extents()` (3rd line);
  - `get_iter()` (1st line)

if i comment these function calls out (e.g. `pango.pango_layout_set_markup()`), no segfault occurs (but of course the tests still fail because now all these functions return incorrect values). i tried commenting out `_GC_METHOD` to see if disabling GC would do something, but it didn't.

leifgehrmann commented:
> i noticed the weird error when copying a Color, but i forgot to mention it. it does seem like `pango_color_copy` might be broken, because i remember testing the FFI function directly and the object i got back was wrong. i'll try doing some more experimenting though. if it somehow does end up being broken we could just manually copy the RGB attributes to a new pointer? (and file an issue with pango!).

Yeah, I think manually creating a new `Color` object and copying the values is a reasonable solution for the short term. I looked at Pango's source code and I do see that [they have a test for this](https://gitlab.gnome.org/GNOME/pango/-/blob/4276593e8aa4b7a84ef504df1525865c7ba2fc8d/tests/testcolor.c#L135), so it's a little strange for this to happen. I'll investigate it later and raise an issue with Pango if there is an issue.

> i also see another test case that asserts that a NotImplementedError is raised in `Color(...) == 1`, i assume i.e. whenever a Color is compared to a non-object... i had changed `PangoObject#__eq__()` so that it wouldn't error and simply return False; should i revert the change or remove the failing test case?

Yeah, I think you are right, returning false should be fine. Just update the test to check `Color(...) == 1` is `False`, that way code coverage is maintained.

leifgehrmann commented on pangocffi/color.py:
```diff
@@ -1,102 +1,77 @@
+class Color(PangoObject):
+    _INIT_METHOD = ffi.new
+    _INIT_CLASS = "PangoColor"
+    _GC_METHOD = pango.pango_color_free
```
Removing this fixes the SegFault issues

```suggestion
```

leifgehrmann commented on pangocffi/color.py:
```diff
@@ -107,28 +82,26 @@ def parse_color(self, spec: str) -> bool:
 
-    def __eq__(self, col: "Color") -> bool:
-        if isinstance(col, Color):
+    # TODO: remove this and go back to using _COPY_METHOD
+    def __copy__(self) -> "Color":
+        return Color(self.red, self.green, self.blue)
```
To get the tests working, we also need to override the `copy()` method itself.

I've also updated the comment to point to a new issue I created here: https://github.com/leifgehrmann/pangocffi/issues/49

```suggestion
    # TODO: remove this and go back to using _COPY_METHOD
    # See https://github.com/leifgehrmann/pangocffi/issues/49
    def copy(self) -> "Color":
        return self.__copy__()

    def __copy__(self) -> "Color":
        return Color(self.red, self.green, self.blue)
```

leifgehrmann commented on pangocffi/glyph_item_iter.py:
```diff
@@ -24,23 +24,10 @@ class GlyphItemIter:
-        return self._pointer
+    _INIT_METHOD = ffi.new
+    _INIT_CLASS = "PangoGlyphItemIter"
+    _GC_METHOD = pango.pango_glyph_item_iter_free
```
Removing this fixes the SegFault issues.

```suggestion
```

leifgehrmann commented on tests/functional/test_color.py:
```diff
@@ -36,5 +36,6 @@ def test_context_returns_null_from_null_pointer(self):
-            assert a == 1
+        b = Color(1, 1, 1)
+        assert a != b
+        assert a != 1
```
Increase the code coverage a bit by testing the `from_pointer` method.

```suggestion
        assert a != 1
        
    def test_from_pointer(self):
        a = Color(1, 2, 3)
        b = Color.from_pointer(a.pointer)
        assert a == b
```

leifgehrmann commented on pangocffi/layout.py:
```diff
@@ -15,268 +15,165 @@ class Layout(object):
-        """
-        Returns the alignment for the layout: how partial lines are positioned
-        within the horizontal space available.
+    aligmment: Alignment = property(_get_alignment, _set_alignment)
```
Typo 😄

```suggestion
    alignment: Alignment = property(_get_alignment, _set_alignment)
```

leifgehrmann commented on tests/functional/test_layout.py:
```diff
@@ -39,47 +39,47 @@ def test_layout_setting_font_description():
-        layout.set_attributes(attr_list)
-        layout.get_attributes()
+        layout.attributes = attr_list
+        layout.attributes
```
This was originally my fault when I wrote the test, but there should be an assertion here.

```suggestion
        assert layout.attributes is not None
```

leifgehrmann commented on tests/functional/test_layout.py:
```diff
@@ -39,47 +39,47 @@ def test_layout_setting_font_description():
-        layout.set_attributes(None)
-        layout.get_attributes()
+        layout.attributes = None
+        layout.attributes
```
```suggestion
        assert layout.attributes is None
```

leifgehrmann commented on tests/functional/test_layout.py:
```diff
@@ -39,47 +39,47 @@ def test_layout_setting_font_description():
-        layout.get_attributes()
+        layout.attributes = attr_list
+        layout.attributes
 
```
Add a test here for `_del_attributes`

```suggestion

        del layout.attributes
        assert layout.attributes is None
```

leifgehrmann reviewed (commented):
Hi there! I spent some time trying to find the SegFault error and I believe I found the issue.

Basically it appears the `_GC_METHOD` for `color.py` and `glyph_item_iter.py` are not required. Don't ask me why it's not required, but that's the way I set it up originally in master apparently.

I've also left some comments suggesting some changes which should fix some of the other tests and also increase code coverage a tiny bit.

There are also a few lines of code that need changing, but annoyingly GitHub doesn't allow me to comment on them directly. So here's a list:

* In `test_attributes.py` modify line 75 to read `markup_texts = ('pango.apply_markup(\'...\')', 'Attribute.from_shape()')` (i.e. s/set_markup/apply_markup)
* In `test_attr_list.py` modify line 72 to read `assert a != 2`
* In `test_attribute.py` modify line 263 to read `assert a != 6`

All these changes pass on my machine, so hopefully it'll pass in the GitHub Actions as well.

cAttte commented:
thank you so much for finding the issue! i spent some time messing with the classes and GC methods but it was pretty confusing so i couldn't get it to work, and i haven't had a lot of time during the week.

i'd actually implemented the three small fixes you listed, but completely forgot to commit/push them! also, i fixed the copy method, but directly modified `Color.copy()` instead, since it's already called by `__copy__()` and `__deepcopy__()` up in PangoObject. thanks for mentioning the rest!

i hope everything's alright now, and i'm very sorry for the messy PR and amount of commits. the segfault when testing really hindered my ability to... you know... test the code, so i kind of just trusted myself. i'll make sure to be more careful when contributing further :)

leifgehrmann commented:
Ran the checks again... Looks like one more change is required. (Hopefully 😅)

I think testing `AttrList` for equality by checking the pointers when using older versions of Pango is a bit confusing. `__eq__` raising an AttributeError is fine, and is ultimately more honest to the user. So I'd change it to:

```py
    def __eq__(self, other) -> bool:
        if isinstance(other, PangoObject):
            return bool(
                pango.pango_attr_list_equal(self.pointer, other.pointer)
            )
        return False
```

The alternative is to change the test to assert `False` if Pango <1.46.0 is running, and `True` if Pango >=1.46, but that seems a little hacky in opinion.

cAttte commented:
you're right! while thinking of the concept of `__eq__()` i realized it's kind of "unfair" how some objects get their attributes compared individually, while others simply compare the pointers, which isn't really how it should work at all. i'll change it right now, but it does come to mind that `PangoObject` does the same thing:

https://github.com/cAttte/pangocffi/blob/ab98f56831991dbe89444911e10067f1ce65ae86/pangocffi/pango_object.py#L104-L107

and considering that most classes don't have an `_EQ_METHOD`, this is true basically everywhere. should we override this in every class, find a way to make it declarable, or leave it as is?

leifgehrmann commented:
> but it does come to mind that PangoObject [does the same thing](https://github.com/cAttte/pangocffi/blob/ab98f56831991dbe89444911e10067f1ce65ae86/pangocffi/pango_object.py#L104-L107):

Hmm, I think this can all be improved by flipping the order of operations. In other words, check for pointer equality first, then use the `_EQ_METHOD`. So replace update `PangoObject` to have:

```py
    def __eq__(self, other) -> bool:
        if isinstance(other, PangoObject):
            if self.pointer == other.pointer:
                return True
            if self._EQ_METHOD:
                return bool(self._EQ_METHOD(self.pointer, other.pointer))
        return False
```

and remove the override if `__eq__` in `attr_list.py` and add `_EQ_METHOD = pango.pango_attr_list_equal` at the top.

That way, all PangoObjects correctly do pointer comparisons regardless of whether the object has `_EQ_METHOD`, and if they are different, fallback to the `_EQ_METHOD`. And if the `_EQ_METHOD` fails because of the raised `AttributeError` in old versions of Pango, it's up to the user to either upgrade their version of Pango to get _real_ object comparisons or avoid doing object comparisons in the first place.

----

Other than that, I think this is good for merging. Coverage is going down by a tiny percentage, but that's okay. At the end of the day it's still 99% 😂

cAttte commented:
sure, that makes sense! but won't adding `_EQ_METHOD` back to the top of the file crash outdated versions on start-up again?

leifgehrmann commented:
Yep, you're right, I forgot about that! 🤦

So yeah, keep the override in `attr_list.py`, but reorder the conditionals to check pointer equality first in both implementations of `__eq__()`.

cAttte commented:
coverage is now `-0.01%` :')

leifgehrmann commented:
I think a lower code coverage was always inevitable because you refactored redundant code neatly into `pango_object.py`. 😄

But this all looks great. I'll get this merged in a few minutes (unless you have any last minute additions) and I'll prepare a new release today if possible.

I'll also be updating `pangocairocffi` soon as well (Got a PR prepared here: https://github.com/leifgehrmann/pangocairocffi/pull/29)

Thanks again for this contribution, I really appreciate the amount of work that went the refactoring all the objects into `PangoObject`!

cAttte commented:
thank you too! this was a nice experience and i'll probably make a PR to implement more classes from the Pango API soon. :)
